# Antibody-Antigen ΔΔG Prediction

This repository provides code for predicting changes in binding affinity (ΔΔG) upon point mutations to antibody or antigen sequences, using regression models built on top of pretrained protein language models.

## Installation

```bash
conda env create -f environment.yml
```

## Datasets

We use the synthetic ddG dataset created by Hummer et al., which was generated by running Rosetta Flex ddG on 1302 antibody-antigen complexes from SabDab with 16 mutants per complex for a total of 20,829 variants. The synthetic ddG labels for each mutant complex can be downloaded from [https://github.com/oxpig/Graphinity/tree/main/data/ddg_synthetic/Flex_ddG](https://github.com/oxpig/Graphinity/tree/main/data/ddg_synthetic/Flex_ddG).

The associated PDB files were extracted from the `synthetic_flexddg_ddg_wt_pdbs.tar.gz` file in [https://doi.org/10.5281/zenodo.15384945](https://doi.org/10.5281/zenodo.15384945).

### Dataset Preparation

Extract the antibody heavy chain, antibody light chain, and antigen sequences from the PDB files. This script creates a JSON file where each key is a chain in each mutant complex with the following data: 
- `sequence`: The amino acid sequence of the chain
 - `mutation_index`: The position in the sequence where a mutation occurs (only included for mutated chains)

```bash
python scripts/convert_pdb_to_fasta.py \
    --pdb_dir ddg_synthetic/Flex_ddG/flexddg_wt_pdbs \
    --save_path ddg_synthetic/Flex_ddG/flexddg_wt_pdbs.json
```

Read from the JSON file and create a CSV file combining the ddG labels and the 2 antibody chain and 1 antigen chain sequences. If the complex only contains 1 antibody chain, the empty string is stored in the second antibody chain.

```bash
python scripts/build_flexddg_dataset.py \
    --json_path ddg_synthetic/Flex_ddG/flexddg_wt_pdbs.json \
    --ddg_path ddg_synthetic/Flex_ddG/Synthetic_FlexddG_ddG_20829.csv \
    --save_path ddg_synthetic/Flex_ddG/Synthetic_FlexddG_ddG_20829_with_sequences.csv
```

## PLM Embeddings

Use the ESM2 protein language model to embed the sequences of the antibody H chain, antibody L chain, and the antigen chain.

```bash
python scripts/esm_embedding.py \
    --esm_model esm2_t33_650M_UR50D \
    --last_layer 33 \
    --save_path ddg_synthetic/Flex_ddG/embeddings/embeddings_flexddg.pt \
    --average_embeddings \
    --sequences_path ddg_synthetic/Flex_ddG/Synthetic_FlexddG_ddG_20829_with_sequences.csv
```
